networking/src/p2p/peer/quota.rs

```diff
@@ -0,0 +1,202 @@
+                        default[i] = (tx, rx);
+                    }
+                }
+            }
+        }
+        default
+    };
+}
+
+pub(crate) struct ThrottleQuota {
+    quotas: [(isize, isize); THROTTLING_QUOTA_NUM],
+    quota_disabled: bool,
+    log: Logger,
+}
+
+impl ThrottleQuota {
+    pub(crate) fn new(log: Logger) -> Self {
+        Self {
+            quotas: THROTTLING_QUOTA_MAX.clone(),
+            quota_disabled: *THROTTLING_QUOTA_DISABLE,
+            log,
+        }
+    }
+
+    pub(crate) fn msg_index(msg: &PeerMessageResponse) -> usize {
+        debug_assert!(THROTTLING_QUOTA_NUM == 18);
+        match msg.message() {
+            PeerMessage::Disconnect => 0,
+            PeerMessage::Advertise(_) => 1,
+            PeerMessage::SwapRequest(_) => 2,
+            PeerMessage::SwapAck(_) => 3,
+            PeerMessage::Bootstrap => 4,
+            PeerMessage::GetCurrentBranch(_) => 5,
+            PeerMessage::CurrentBranch(_) => 6,
+            PeerMessage::Deactivate(_) => 7,
+            PeerMessage::GetCurrentHead(_) => 8,
+            PeerMessage::CurrentHead(_) => 9,
+            PeerMessage::GetBlockHeaders(_) => 10,
+            PeerMessage::BlockHeader(_) => 11,
+            PeerMessage::GetOperations(_) => 12,
+            PeerMessage::Operation(_) => 13,
+            PeerMessage::GetProtocols(_) => 14,
+            PeerMessage::Protocol(_) => 15,
+            PeerMessage::GetOperationsForBlocks(_) => 16,
+            PeerMessage::OperationsForBlocks(_) => 17,
+        }
+    }
+
+    fn index_to_str(index: usize) -> &'static str {
+        if index < THROTTLING_QUOTA_NUM {
+            THROTTLING_QUOTA_STRS[index]
+        } else {
+            "<invalid index>"
+        }
+    }
+
+    pub fn get_message_kind(msg: &PeerMessageResponse) -> &'static str {
+        Self::index_to_str(Self::msg_index(msg))
+    }
+
+    pub fn can_send(&mut self, msg: &PeerMessageResponse) -> Result<(), isize> {
+        let index = Self::msg_index(msg);
+        if THROTTLING_QUOTA_MAX[index].0 <= 0 {
+            return Ok(());
+        }
+        self.quotas[index].0 = self.quotas[index].0.checked_sub(1).ok_or(self.quotas[index].0)?;
+        if self.quotas[index].0 >= 0 {
+            self.quotas[index].1 =
+                cmp::min(self.quotas[index].1 + 1, THROTTLING_QUOTA_MAX[index].1);
+            Ok(())
+        } else if self.quota_disabled {
+            Ok(())
+        } else {
+            Err(self.quotas[index].0)
+        }
+    }
+
+    pub fn can_receive(&mut self, msg: &PeerMessageResponse) -> Result<(), isize> {
+        let index = Self::msg_index(msg);
+        if THROTTLING_QUOTA_MAX[index].1 <= 0 {
+            return Ok(());
+        }
+        self.quotas[index].1 = self.quotas[index].0.checked_sub(1).ok_or(self.quotas[index].1)?;
+        if self.quotas[index].1 >= 0 {
+            self.quotas[index].0 =
+                cmp::min(self.quotas[index].0 + 1, THROTTLING_QUOTA_MAX[index].0);
```

Comment:
@akoptelov 
what about this `self.quotas[index].0 + 1`  if anybody will set quota to i32::MAX it will trigger overflow in debug mode I guess, better use checked_add

and also later `"amount" => -self.quotas[index].0 + tx_max`